builtins/src/main/java/org/jline/builtins/PosixCommands.java

```diff
@@ -388,24 +388,18 @@ public static void cat(Context context, String[] argv) throws Exception {
         if (args.isEmpty()) {
             args = Collections.singletonList("-");
         }
-        List<InputStream> expanded = new ArrayList<>();
+        List<Source> sources = new ArrayList<>();
         for (String arg : args) {
             if ("-".equals(arg)) {
-                expanded.add(context.in());
+                sources.add(new StdInSource(context.in()));
             } else {
                 maybeExpandGlob(context, arg)
-                        .map(p -> {
-                            try {
-                                return p.toUri().toURL().openStream();
-                            } catch (IOException e) {
-                                throw new RuntimeException(e);
-                            }
-                        })
-                        .forEach(expanded::add);
+                        .map(p -> new PathSource(p, p.toString()))
+                        .forEach(sources::add);
```

Comment:
This whole `if ("-".equals(arg)) { ... } else { maybeExpandGlob(...)` can be refactored in a common method.  I've tried to do that in #1407 , but maybe too many changes in that one...

Reply:
Groovy has that refactored here:

https://github.com/apache/groovy/blob/master/subprojects/groovy-groovysh/src/main/groovy/org/apache/groovy/groovysh/jline/GroovyPosixCommands.java#L1131-L1145

That picks up "-", globs/files and $variables. I am not sure the variables is done the best way - it is done by passing through a String[]. But it means the usual pipe style variations work for all the Posix Commands, e.g.:

```
files=/ls --1
head=/head $files -n10
/tail $head -n5
```

Or:

```
/ls --1 |; /head -n10 |; /tail -n5
```